lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,208 @@
+using namespace llvm;
+
+namespace {
+
+class SPIRVLowerBitCastToNonStandardTypeBase {
+public:
+  SPIRVLowerBitCastToNonStandardTypeBase() {}
+
+  bool runLowerBitCastToNonStandardType(Module &Module) {
+    // This pass doesn't cover all possible uses of non-standard types, only
+    // known
+    auto *M = &Module;
+    bool Changed = false;
+
+    std::vector<std::pair<Instruction *, VectorType *>> BCastsToNonStdVec;
+    std::stack<Instruction *> InstsToErase;
+    for (auto &F : M->functions())
+      for (auto &BB : F)
+        for (auto &I : BB)
+          if (BitCastInst *BC = dyn_cast<BitCastInst>(&I)) {
+            auto *DestTy = BC->getDestTy();
+            if (auto *ElemTy = dyn_cast<PointerType>(DestTy))
+              DestTy = ElemTy->getElementType();
+            if (auto *DestVecTy = dyn_cast<VectorType>(DestTy)) {
+              uint64_t NumElemsInDestVec =
+                  DestVecTy->getElementCount().getValue();
+              if (NumElemsInDestVec != 2 && NumElemsInDestVec != 3 &&
+                  NumElemsInDestVec != 4 && NumElemsInDestVec != 8 &&
+                  NumElemsInDestVec != 16)
+                BCastsToNonStdVec.push_back(std::make_pair(&I, DestVecTy));
+            }
+          }
+
+    for (auto &I : BCastsToNonStdVec)
+      Changed |= lowerBitCastToNonStdVec(I, InstsToErase);
+
+    while (!InstsToErase.empty()) {
+      InstsToErase.top()->eraseFromParent();
+      InstsToErase.pop();
+    }
+
+    verifyRegularizationPass(*M, "SPIRVLowerBitCastToNonStandardType");
+    return Changed;
+  }
+
+  bool lowerBitCastToNonStdVec(std::pair<Instruction *, VectorType *> &Inst,
+                               std::stack<Instruction *> &InstsToErase) {
+    bool Changed = false;
+    std::vector<Instruction *> InstsToInsert;
+    Instruction *I = Inst.first;
+    VectorType *DestVecTy = Inst.second;
+    uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+    InstsToErase.push(I);
+    AddrSpaceCastInst *Src{nullptr};
+    auto *II = I;
+    while ((II = II->getNextNode())) {
+      if (auto *ASCI = dyn_cast<AddrSpaceCastInst>(II)) {
+        unsigned DestAddrSpace = ASCI->getDestAddressSpace();
+        auto *SrcPointer = cast<CastInst>(I)->getSrcTy();
+        auto *SrcTy = cast<PointerType>(SrcPointer)->getElementType();
+        SrcPointer = SrcTy->getPointerTo(DestAddrSpace);
+        if (ASCI->getOperand(0) == cast<Value>(I)) {
+          Src = new AddrSpaceCastInst(I->getOperand(0), SrcPointer);
+          InstsToInsert.push_back(Src);
+          InstsToErase.push(II);
+          break;
+        }
+      }
+    }
+    if (!II)
+      II = I;
+    auto *SrcTy =
+        cast<PointerType>(cast<Value>(Src)->getType())->getElementType();
+    LoadInst *Load =
+        Src ? new LoadInst(SrcTy, Src, "", false, Align())
+            : new LoadInst(SrcTy, I->getOperand(0), "", false, Align());
+    InstsToInsert.push_back(Load);
+    auto *III = II;
+    while ((III = III->getNextNode())) {
```

Comment:
This pass seems to handle one hardcoded pattern, right? Why don't we go through users of faulty bitcast instead (and users of users until the bad type is gone)? I think it can make code simpler (at least I assume you won't need that much IIIs)